bccsp/sw/keyimport.go

```diff
@@ -123,12 +124,16 @@ func (ki *x509PublicKeyImportOptsKeyImporter) KeyImport(raw interface{}, opts bc
 
 	pk := x509Cert.PublicKey
 
-	switch pk.(type) {
+	switch pk := pk.(type) {
 	case *ecdsa.PublicKey:
 		return ki.bccsp.KeyImporters[reflect.TypeOf(&bccsp.ECDSAGoPublicKeyImportOpts{})].KeyImport(
 			pk,
 			&bccsp.ECDSAGoPublicKeyImportOpts{Temporary: opts.Ephemeral()})
+	case *rsa.PublicKey:
+		// This path only exists to support environments that use RSA certificate
+		// authorities to issue ECDSA certificates.
+		return &rsaPublicKey{pubKey: pk}, nil
 	default:
-		return nil, errors.New("Certificate's public key type not recognized. Supported keys: [ECDSA]")
+		return nil, errors.New("Certificate's public key type not recognized. Supported keys: [ECDSA, RSA]")
```

Comment:
This makes me wonder - let's say I have a v2.2 Fabric environment prior to this commit and I add an RSA CA to my config block. What happens? Does the peer crash when it initializes an MSP manager? Or does the peer simply invalidate transactions signed by identities issued by the RSA CA? If it's the former then we're safe to add this, but if it's the latter it is some potential for non-determinism across versions.

Reply:
I had a discussion with @jyellick about this. My takeaway was that the failure mode presents as crashing peers and orderers. 

In the case of the peer, if an active configuration is using an RSA CA and the peer is missing the RSA code (2.x), starting a peer that does not support it results in a crash. If a config block arrives with an RSA CA, validation fails and the peer panics.

There were subtleties related to validation performed by the orderer running in a mixed version cluster that I didn't completely grok so he may need to comment on those scenarios.

At the end of the conversation, we did come away concluding that this would not result in non-determinism.

Reply:
To add a little more context to what @sykesm has said above, in this case, because we are dealing only with CA certs and not with signing certs a lot of the complications disappear.  The only transactions which depend on CA cert evaluations are channel config updates.

Channel config updates are always preprocessed by the orderer.  In general, the update is first checked by the orderer to which it was Broadcast.  Then there may be a further check by the leader in the Raft case.  If either of these checks fails, then the config update is simply dropped, and it does not appear in the chain.  If both of these checks pass, then the config update is applied as a new config block in the chain.  The important thing to note is that config correctness checks always occur _before_ consensus.  This was a critical change that was made between v1.0 and v1.1 which practically eliminates the possibility of forking the blockchain at the orderer based on non-deterministic config update processing.

Once a config change achieves order via consensus, it is simply applied, with no validation step.  If the apply fails, then the node simply crashes.

At the peer side, the behavior is similar, though slightly different.  The peer does not trust that the ordering service appropriately validated the config update, so the peer redoes the signature validation etc. before finally applying the new config.  If the signature checking fails, or applying the new config fails, the peer, like the orderer, simply crashes.

So, for any config update, which is non-deterministically evaluated, we still have the guarantee that it either applies everywhere or not at all.  In a mixed mode network, with v1.4.x and v2.1.x nodes, the v2.1.x orderers would reject the config update, the v1.4.x ones would accept it.  So, if you submitted to a v1.4.x node, and the leader for the channel was a v1.4.x node, all the v2.1.x orderers would crash, but, there would be no fork.  If either the node receiving the broadcast or the leader were v2.1.x then the config update would simply be rejected or dropped.

TLDR; Matt is absolutely right here, there is no potential to fork.  Although, it's possible for a malicious config update to get through in a mixed mode network with binaries that do and don't support RSA TLS CAs, this would simply result in a crash, which could be fixed by moving to a newer version of Fabric.

Reply:
Got it, thanks Matt and Jason.